### PR TITLE
http health check: configurable expected HTTP response codes

### LINKED ISSUE #4367: healthcheck: Allow non-200 status codes as healthy HTTP healthchecks
*Title*: *healthcheck: Allow non-200 status codes as HTTP healthchecks*

*Description*:
We have certain services that do not return a 200 on a successful healthcheck. The most common one I've seen is returning an HTTP 204. Envoy marks these as unhealthy because it is checking [explicitly](https://github.com/envoyproxy/envoy/blob/master/source/common/upstream/health_checker_impl.cc#L180) for a 200

I think a good default would be 2xx or 3xx is considered healthy. Whether a given status code should be considered healthy is up for debate, but i think this is pretty standard for proxies right now, and so i think it makes sense to do this.

For extra credit, we can have configuration for what constitutes healthy. e.g. regex check on the body of the response.

### PR DESCRIPTION
**Pull request by bartebor**
*Description:* Makes it possible for http health checks to accept any http response codes configured in a list of ranges `expected_statuses`. The default behaviour is unchanged - only HTTP 200 response is considered healthy.
*Risk Level*: Low (does not modify default behaviour)
*Testing*: Unit tests
*Docs Changes*: comments describing new field in proto file, short note in health_checking.rst
*Release Notes*: added info about expected responses configuration
Fixes #4367

### CODE REVIEW

File: source/common/upstream/http_status_checker.cc [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    if (end > 600) {
+      throw EnvoyException(
+          fmt::format("Invalid http status range: expecting end <= 600, but found end={}", end));
+    }
+
+    httpStatusChecker->ranges_.emplace_back(
+        std::make_pair(static_cast<uint64_t>(start), static_cast<uint64_t>(end)));
+  }
+
+  if (httpStatusChecker->ranges_.empty()) {
```
Comment (venilnoronha) [inline source/common/upstream/http_status_checker.cc]: Should we keep `200` irrespective of the input ranges?

Comment (bartebor) [reply source/common/upstream/http_status_checker.cc]: In my opinion it is better to give users full control of accepted responses. Users will probably just extend the range using [200, 300) or [200, 400), so 200s will be included explicitly. I don't know why someone would need to disable 200s, but if they wanted, they could.

Comment (venilnoronha) [reply source/common/upstream/http_status_checker.cc]: Can we explictly mention this behavior in the proto (or elsewhere)? Thanks.

Comment (bartebor) [reply source/common/upstream/http_status_checker.cc]: I clarified comment in proto file, it should be clear now.

File: source/common/upstream/http_status_checker.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+#include <vector>
+
+#include "envoy/api/v2/core/health_check.pb.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Upstream {
+
+class HttpStatusChecker;
```
Comment (junr03) [inline source/common/upstream/http_status_checker.h]: overall comment about this class: should we have this as a class in its own file, or given that the only use is in the http health checker, should we make it a struct within the http health checker for now?

Comment (junr03) [reply source/common/upstream/http_status_checker.h]: There are several places in the codebase were we follow that pattern for small abstractions that only relate to one class.

File: api/envoy/api/v2/core/health_check.proto [hunk 3]
```diff
@@ -119,6 +120,10 @@ message HealthCheck {
 
     // If set, health checks will be made using http/2.
     bool use_http2 = 7;
+
+    // Specifies a list of HTTP response statuses considered healthy. If provided, replaces default
+    // 200-only policy - 200 must be included explicitly as needed.
```
[the remark this answers is not in the document]
Comment (bartebor) [reply api/envoy/api/v2/core/health_check.proto]: This is already documented in [Int64Range](https://www.envoyproxy.io/docs/envoy/latest/api-v2/type/range.proto.html#type-int64range). Would you like to repeat it here?

File: source/common/upstream/http_status_checker.h (hunk 2 above)
Comment (bartebor) [reply source/common/upstream/http_status_checker.h]: I generally prefer smaller files with less responsibility, separate testing etc. so I made it this way.
As this is my first PR in this project I am not aware of general patterns you use. I can refactor this into inner class if you decide so.

File: source/common/upstream/http_status_checker.h [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+/**
+ * Utility class checking if given http status matches configured expectations.
+ */
+class HttpStatusChecker {
+public:
+  static HttpStatusCheckerPtr
+  configure(const Protobuf::RepeatedPtrField<envoy::type::Int64Range>& expected_statuses,
+            uint64_t default_expected_status);
+
+  bool isExpected(uint64_t http_status) const;
```
[the remark this answers is not in the document]
Comment (bartebor) [reply source/common/upstream/http_status_checker.h]: I used similar name in proto, would you like to change it to something like 'status_ranges' too?

**Comment by bartebor** [comment]
@junr03 I have made changes you requested.

File: api/envoy/api/v2/core/health_check.proto (hunk 3 above)
[the remark this answers is not in the document]
Comment (junr03) [reply api/envoy/api/v2/core/health_check.proto]: my opinion is that not everyone will go over to that documentation, so a reminder here helps. But, you could argue that is duplication. Up to you!

[the remark this answers is not in the document]
Comment (bartebor) [reply api/envoy/api/v2/core/health_check.proto]: I mentioned that is follows `Int64Range` semantics.

**Comment by mattklein123** [comment]
@alyssawilk is OOO this week, I'm happy to take a look.

File: source/common/upstream/health_checker_impl.h [hunk 5]
```diff
@@ -121,6 +135,7 @@ class HttpHealthCheckerImpl : public HealthCheckerImplBase {
   const std::string host_value_;
   absl::optional<std::string> service_name_;
   Router::HeaderParserPtr request_headers_parser_;
+  const std::unique_ptr<HttpStatusChecker> http_status_checker_;
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.h]: nit: This doesn't actually need to be an allocated pointer. You can just make this a const inline member variable and just initialize it as part of the initializer list in the constructor?

File: source/common/upstream/health_checker_impl.cc [hunk 6]
```diff
@@ -94,12 +94,55 @@ HttpHealthCheckerImpl::HttpHealthCheckerImpl(const Cluster& cluster,
...
   if (!config.http_health_check().service_name().empty()) {
     service_name_ = config.http_health_check().service_name();
   }
 }
 
+HttpHealthCheckerImpl::HttpStatusChecker::HttpStatusChecker(
+    const Protobuf::RepeatedPtrField<envoy::type::Int64Range>& expected_statuses,
+    uint64_t default_expected_status) {
+  for (const auto& status_range : expected_statuses) {
+    auto start = status_range.start();
```
Comment (mattklein123) [inline source/common/upstream/health_checker_impl.cc]: nit: `const auto`, same next line.

File: source/common/upstream/health_checker_impl.h (hunk 5 above)
Comment (bartebor) [reply source/common/upstream/health_checker_impl.h]: Pointer removed.

File: source/common/upstream/health_checker_impl.cc (hunk 6 above)
Comment (bartebor) [reply source/common/upstream/health_checker_impl.cc]: Done!

**Pull request merged**